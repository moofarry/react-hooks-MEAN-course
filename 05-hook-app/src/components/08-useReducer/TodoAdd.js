import React from "react";
import { useForm } from "../../hooks/useForm";

export const TodoAdd = ({ handleAddTodo }) => {
  const [{ description }, handleInputChange, reset] = useForm({
    description: "",
  });
  const handleSubmit = (e) => {
    e.preventDefault();
    if (description.trim() <= 1) {
      return;
    }

    const newTodo = {
      id: new Date().getTime(),
      description: description,
      done: false,
    };
    handleAddTodo(newTodo);
    reset();
  };
  return (
    <>
      <h4>Agregar TODO</h4>
      <hr />
      <form onSubmit={handleSubmit}>
        <input
          className="form-control"
          type="text"
          name="description"
          placeholder="Add new TODO"
          autoComplete="off"
          onChange={handleInputChange}
          value={description}
        />
        <button className="btn btn-outline-primary mt-1 w-100 " type="submit">
          Agregar
        </button>
      </form>
    </>
  );
};
