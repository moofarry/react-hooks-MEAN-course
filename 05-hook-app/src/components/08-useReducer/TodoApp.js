import React, { useReducer, useEffect } from "react";
import { todoReducer } from "./todoReducer";
import "./styles.css";
import { useForm } from "../../hooks/useForm";
import { TodoList } from "./TodoList";

const init = () => {
  return JSON.parse(localStorage.getItem("todos")) || [];
};

export const TodoApp = () => {
  const [todos, dispatch] = useReducer(todoReducer, [], init);
  const [{ description }, handleInputChange, reset] = useForm({
    description: "",
  });

  useEffect(() => {
    localStorage.setItem("todos", JSON.stringify(todos));
  }, [todos]);

  const handleSubmit = (e) => {
    e.preventDefault();

    if (description.trim() <= 1) {
      return;
    }
    const newTodo = {
      id: new Date().getTime(),
      description: description,
      done: false,
    };
    const action = {
      type: "add",
      payload: newTodo,
    };
    dispatch(action);
    reset();
  };

  const handleRemove = (todoId) => {
    const action = {
      type: "remove",
      payload: todoId,
    };
    dispatch(action);
  };

  const handleToggle = (todoId) => {
    dispatch({
      type: "toggle",
      payload: todoId,
    });
  };

  return (
    <div>
      <h1>Cosas por hacer morchis : ({todos.length})</h1>
      <hr />
      <div className="row">
        <div className="col-6 me-5">
          <TodoList
            todos={todos}
            handleToggle={handleToggle}
            handleRemove={handleRemove}
          />
        </div>
        <div className="col-4">
          <h4>Agregar TODO</h4>
          <hr />
          <form onSubmit={handleSubmit}>
            <input
              className="form-control"
              type="text"
              name="description"
              placeholder="Add new TODO"
              autoComplete="off"
              onChange={handleInputChange}
              value={description}
            />
            <button
              className="btn btn-outline-primary mt-1 w-100 "
              type="submit"
            >
              Agregar
            </button>
          </form>
        </div>
      </div>
    </div>
  );
};
