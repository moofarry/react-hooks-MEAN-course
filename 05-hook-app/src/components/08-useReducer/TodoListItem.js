import React from "react";

export const TodoListItem = ({ todo, index, handleRemove, handleToggle }) => {
  return (
    <li key={todo.id} className="list-group-item">
      <p
        className={`${todo.done && "complete"}`}
        onClick={() => handleToggle(todo.id)}
      >
        {index + 1}. {todo.description}
      </p>
      <button className="btn btn-danger" onClick={() => handleRemove(todo.id)}>
        Borrar
      </button>
    </li>
  );
};
