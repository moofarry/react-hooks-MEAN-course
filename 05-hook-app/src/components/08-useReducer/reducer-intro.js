const initialState = [
  {
    id: 1,
    todo: "Go to school",
    done: false,
  },
];

const todoReducer = (state = initialState, action) => {
  if (action?.type === "add") {
    return [...state, action.payload];
  }

  return state;
};

let todos = todoReducer();

const newTodo = {
  id: 2,
  todo: "Do homework",
  done: false,
};

const action = {
  type: "add",
  payload: newTodo,
};

todos = todoReducer(todos, action);

console.log(todos);
