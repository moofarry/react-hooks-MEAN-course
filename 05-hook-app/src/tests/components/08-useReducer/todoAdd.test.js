import React from "react";
import { shallow } from "enzyme";
import { TodoAdd } from "../../../components/08-useReducer/TodoAdd";

describe("test in <TodoAdd/>", () => {
  const handleAddTodo = jest.fn();

  const wrapper = shallow(<TodoAdd handleAddTodo={handleAddTodo} />);
  test("should show correctly", () => {
    expect(wrapper).toMatchSnapshot();
  });

  test("should dont call addTodo", () => {
    const formSubmit = wrapper.find("form").prop("onSubmit");
    formSubmit({ preventDefault() {} });

    expect(handleAddTodo).toHaveBeenCalledTimes(0);
  });

  test("should call handleAddTodo", () => {
    wrapper.find("button").simulate("click");

    const value = "Learn React";
    wrapper.find("input").simulate("change", {
      target: {
        value: value,
        name: "description",
      },
    });
    const formSubmit = wrapper.find("form").prop("onSubmit");

    formSubmit({ preventDefault() {} });
    expect(handleAddTodo).toHaveBeenCalledTimes(1);
    expect(handleAddTodo).toHaveBeenCalledWith({
      id: expect.any(Number),
      description: value,
      done: false,
    });
    expect(wrapper.find("input").prop("value")).toBe("");
  });
});
