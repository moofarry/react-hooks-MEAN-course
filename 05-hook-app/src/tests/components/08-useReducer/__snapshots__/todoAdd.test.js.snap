// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test in <TodoAdd/> should show correctly 1`] = `
<Fragment>
  <h4>
    Agregar TODO
  </h4>
  <hr />
  <form
    onSubmit={[Function]}
  >
    <input
      autoComplete="off"
      className="form-control"
      name="description"
      onChange={[Function]}
      placeholder="Add new TODO"
      type="text"
      value=""
    />
    <button
      className="btn btn-outline-primary mt-1 w-100 "
      type="submit"
    >
      Agregar
    </button>
  </form>
</Fragment>
`;
